src/_pytest/recwarn.py

```diff
@@ -232,7 +235,7 @@ def __init__(
         self,
         expected_warning: Optional[
             Union[Type[Warning], Tuple[Type[Warning], ...]]
```

Comment:
```suggestion
        expected_warning: Union[
            Type[Warning], Tuple[Type[Warning], ...]
```

Reply:
I decided to keep the `Optional` type, because removing it would break `mypy` for existing uses of `pytest.warns(None)`, like here: https://github.com/pytest-dev/pytest/blob/6ae71a2c2b7bdadf4f16c0eac960838315b51198/testing/test_tmpdir.py#L406

Reply:
My intention was that mypy should fail on the deprecated way - to give people early feedback!

Reply:
The side-effect is that we are getting these errors from mypy ourselves. For the moment I have added a `# type: ignore[call-overload]` comment in the places we are still testing the behaviour of `warns(None)`. Let me know if you think we 'd better address it in a different way.